### PR TITLE
webdav: added a "rclone" WebDAV vendor

### LINKED ISSUE #7160: Add rclone vendor for webdav for use with rclone serve webdav
#### What is your current rclone version (output from `rclone version`)?
rclone 1.63.1

#### What problem are you are trying to solve?
Add a `rclone` vendor to list of webdav vendors. This will allow modified times, hashes and other improvements that the nextcloud vendor previously provided & potentially future extensions to work when using rclone with a webdav remote connecting to a webdav server hosted by `rclone serve`

#### How do you think rclone should be changed to solve that?
For now add a `rclone` vendor to list of vendors, then allow use of those extended features (modify times, hashes, etc) using that vendor rather than only nextcloud/owncloud/fastmail

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

### PR DESCRIPTION
**Pull request by akumar-xyz**
Fixes #7160

#### What is the purpose of this change?

This change adds `rclone-serve` as one of the vendors for WebDAV backend. Only the following attributes are set:
- canStream = `true`
- useOCMtime = `true`

#### Was the change discussed in an issue or in the forum before?

Briefly discussed in github issue [#7160 ](https://github.com/rclone/rclone/issues/7160)

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: backend/webdav/webdav.go [hunk 1]
```diff
@@ -91,6 +91,9 @@ func init() {
 			}, {
 				Value: "sharepoint-ntlm",
 				Help:  "Sharepoint with NTLM authentication, usually self-hosted or on-premises",
+			}, {
+				Value: "rclone-serve",
```
Comment (ncw) [inline backend/webdav/webdav.go:95]: I think just call this `rclone` instead of `rclone-serve`.

File: backend/webdav/webdav.go [hunk 2]
```diff
@@ -644,6 +647,10 @@ func (f *Fs) setQuirks(ctx context.Context, vendor string) error {
 		// so we must perform an extra check to detect this
 		// condition and return a proper error code.
 		f.checkBeforePurge = true
+	case "rclone-serve":
+		f.canStream = true
```
Comment (ncw) [inline backend/webdav/webdav.go:651]: I had a check in the source and I think these are correct.

**Review by ncw — COMMENTED** [review]
To make sure this tests properly we should adjust the unit and integration tests to have the vendor in.

So this needs changing

https://github.com/rclone/rclone/blob/52d444f4a91cc7afd1552574dced484fd3b5138b/cmd/serve/webdav/webdav_test.go#L68

After doing this you should be able to run `go test -v` the unit tests for the server and they still pass.

And this should have a vendor line in

https://github.com/rclone/rclone/blob/52d444f4a91cc7afd1552574dced484fd3b5138b/fstest/testserver/init.d/TestWebdavRclone#L14-L18

This will affect the integration tests (see [test_all -remotes TestWebdavRclone](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#backend-testing))

**Comment by akumar-xyz** [comment]
@ncw thanks for the review, it was very encouraging.
I have made the changes you mentioned and updated the docs as well.

However the tests are failing and I think it's probably because i am running them wrong

Here's what I did
```sh
# start docker
cd backend/webdav
sudo go test -v  # sudo so that the test can start docker
```
and this is what I see
```

    fstest.go:305: 
        	Error Trace:	repo_path/rclone/fstest/fstest.go:305
        	           				repo_path/rclone/fstest/fstest.go:338
        	           				repo_path/rclone/fstest/fstests/fstests.go:696
        	Error:      	Should be true
        	Test:       	TestIntegration/FsMkdir/FsEncoding/invalid_UTF-8
        	Messages:   	listing wrong, want
        	           	 invalid utf-8/invalid utf-8 (100) got
        	           	 invalid utf-8þ/invalid utf-8þ (100)

.
.
.

2023/11/04 15:48:14 NOTICE: TestWebdavOwncloud: Starting server
2023/11/04 15:48:14 NOTICE: Config file "/root/.config/rclone/rclone.conf" not found - using defaults
    fstests.go:447: 
        	Error Trace:	repo_path/rclone/fstest/fstests/fstests.go:447
        	           				repo_path/rclone/backend/webdav/webdav_test.go:28
        	Error:      	Received unexpected error:
        	           	read metadata failed: Propfind "http://127.0.0.1:38081/remote.php/webdav/rclone-test-bisuyux9bikevoy6taguhah6": read tcp 127.0.0.1:53434->127.0.0.1:38081: read: connection reset by peer
        	Test:       	TestIntegration2
        	Messages:   	unexpected error: read metadata failed: Propfind "http://127.0.0.1:38081/remote.php/webdav/rclone-test-bisuyux9bikevoy6taguhah6": read tcp 127.0.0.1:53434->127.0.0.1:38081: read: connection reset by peer
2023/11/04 15:48:29 NOTICE: TestWebdavOwncloud: Stopped server
--- FAIL: TestIntegration2 (315.49s)
=== RUN   TestIntegration3
    fstests.go:423: Using remote "TestWebdavRclone:"
2023/11/04 15:48:30 NOTICE: TestWebdavRclone: Starting server
    fstests.go:433: 
        	Error Trace:	repo_path/rclone/fstest/fstests/fstests.go:433
        	           				repo_path/rclone/backend/webdav/webdav_test.go:42
        	Error:      	Received unexpected error:
        	           	failed to connect to "TestWebdavRclone" on "127.0.0.1:28620"
        	Test:       	TestIntegration3
```

Some guidance will be very appreciated 🙏

**Comment by ncw** [comment]
I pulled your changes locally and tried the tests.

You are right the `go test -v` tests in backend/webdav do fail. I think the failure you saw may be to do with running them with sudo - I don't think that is a good idea. You need to add the user you are using to the docker group so it can control docker directly.

However these aren't run very often though, only when developing the backend. Unlike the full integration tests `test_all -remotes webdav` which are run every day. I ran these and they ran perfectly, so I think your changes are good!

**Comment by akumar-xyz** [comment]
That's fantastic! I'll mark all the tasks as done. Please let me know in case you want me to make any changes here

**Review by ncw — APPROVED** [review]
Thank you - this looks great now. I'll merge :-)

**Pull request merged**